src/contrib/cluster/Akka.Cluster.Sharding/EntityRecoveryStrategy.cs

```diff
@@ -0,0 +1,82 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="PersistentShard.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2016 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Pattern;
+
+namespace Akka.Cluster.Sharding
+{
+    using EntityId = String;
+
+    internal abstract class EntityRecoveryStrategy
+    {
+        public static EntityRecoveryStrategy AllStrategy => new AllAtOnceEntityRecoveryStrategy();
+
+        public static EntityRecoveryStrategy ConstantStrategy(ActorSystem actorSystem, TimeSpan frequency, int numberOfEntities)
+            => new ConstantRateEntityRecoveryStrategy(actorSystem, frequency, numberOfEntities);
+
+        public abstract IImmutableSet<Task<IImmutableSet<EntityId>>> RecoverEntities(IImmutableSet<EntityId> entities);
+    }
+
+    internal class AllAtOnceEntityRecoveryStrategy : EntityRecoveryStrategy
+    {
+        public override IImmutableSet<Task<IImmutableSet<EntityId>>> RecoverEntities(IImmutableSet<EntityId> entities)
+        {
+            return entities.Count == 0
+                ? ImmutableHashSet<Task<IImmutableSet<EntityId>>>.Empty
+                : ImmutableHashSet.Create(Task.FromResult(entities));
+        }
+    }
+
+    internal class ConstantRateEntityRecoveryStrategy : EntityRecoveryStrategy
+    {
+        private readonly ActorSystem actorSystem;
+        private readonly TimeSpan frequency;
+        private readonly int numberOfEntities;
+
+        public ConstantRateEntityRecoveryStrategy(ActorSystem actorSystem, TimeSpan frequency, int numberOfEntities)
+        {
+            this.actorSystem = actorSystem;
+            this.frequency = frequency;
+            this.numberOfEntities = numberOfEntities;
+        }
+
+        public override IImmutableSet<Task<IImmutableSet<EntityId>>> RecoverEntities(IImmutableSet<EntityId> entities)
+        {
+            return entities.Grouped(numberOfEntities).Aggregate(
+                Tuple.Create(frequency, ImmutableHashSet<Task<IImmutableSet<EntityId>>>.Empty),
+                (accumulated, next) => Tuple.Create(accumulated.Item1 + frequency, accumulated.Item2.Add(ScheduleEntities(accumulated.Item1, next.ToImmutableHashSet())))
+            ).Item2;
```

Comment:
I think, that this logic could be simplified. I know, it's direct reflection of what happens on the scala side, but we could aim for clarity here.  Example:

```csharp
var stamp = this.frequency;
var builder = ImmutableHashSet<Task<IImmutableSet<EntityId>>>.Empty.ToBuilder();
foreach (var bucket in entities.Grouped(numberOfEntities)) 
{
    var scheduled = ScheduleEntities(stamp, bucket.ToImmutableHashSet());
    builder.Add(scheduled);
    stamp += this.frequency;
}
return builder.ToImmutable();
```
I think, this is cleaner. It's also a lot less allocations.